libs/gltfio/src/CPUMorpher.cpp

```diff
@@ -0,0 +1,215 @@
+
+CPUMorpher::CPUMorpher(FFilamentAsset* asset) : mAsset(asset) {
+    NodeMap& sourceNodes = asset->mNodeMap;
+    for (auto pair : sourceNodes) {
+        cgltf_node const* node = pair.first;
+        cgltf_mesh const* mesh = node->mesh;
+        if (mesh) {
+            cgltf_primitive const* prims = mesh->primitives;
+            for (cgltf_size pi = 0, count = mesh->primitives_count; pi < count; ++pi) {
+                if (mesh->primitives[pi].targets_count > 0) {
+                    addPrimitive(mesh, pi, &mMorphTable[pair.second]);
+                }
+            }
+        }
+    }
+}
+
+CPUMorpher::~CPUMorpher() {
+    auto engine = mAsset->mEngine;
+    for (auto& entry : mMorphTable) {
+        for (auto& prim : entry.second.primitives) {
+            if (prim.morphBuffer1) engine->destroy(prim.morphBuffer1);
+            if (prim.morphBuffer2) engine->destroy(prim.morphBuffer2);
+        }
+    }
+}
+
+void CPUMorpher::applyWeights(Entity entity, float const* weights, size_t count) noexcept {
+    auto& engine = *mAsset->mEngine;
+    auto renderableManager = &engine.getRenderableManager();
+    auto renderable = renderableManager->getInstance(entity);
+
+    for (auto& prim : mMorphTable[entity].primitives) {
+        size_t size = prim.floatsCount * sizeof(float);
+        float* data = (float*) malloc(size);
+        memset(data, 0, size);
+
+        for (size_t index = 0; index < count; index ++) {
+            const float w = weights[index];
+            if (w < 0.0001f) continue;
+
+            const GltfTarget& target = prim.targets[index];
+            cgltf_size dim = cgltf_num_components(target.type); 
+
+            for (size_t i = 0; i < target.indices.size(); ++i) {
+                uint16_t index = target.indices[i];
+                for (size_t j = 0; j < dim; ++j) {
+                    data[index * dim + j] += w * target.values[i * dim + j];
+                }
+            }
+        }
+        
+        VertexBuffer* vb = prim.vertexBuffer;
+        if (!prim.morphBuffer1) {
+            prim.morphBuffer1 = BufferObject::Builder().size(size).build(engine);
+        }
+        if (!prim.morphBuffer2) {
+            // This is for dealing with a bug in filament shaders where empty normals are not
+            // handled correctly.
+            //
+            // filament shaders deal with tangent frame quaternions instead of normal vectors.
+            // But in case of missing inputs, we get a invalid quaternion (0, 0, 0, 0) instead of
+            // a identity quaterion. This leads to the normals being morphed even no inputs are
+            // given.
+            // 
+            // To fix this, we put a empty morph target at slot 2 and give it a weight of -1.
+            // This won't affect the vertex positions but will cancel out the normal values.
+            //
+            // Note that for this to work, at least two morph targets are required.
```

Comment:
finally figured out the normal morphing bug. We were seeing weird normals before because of this.